src/odb/src/db/dbBlock.cpp

```diff
@@ -1604,6 +1634,76 @@ void dbBlock::addBTermGroup(const std::vector<dbBTerm*>& bterms, bool order)
   block->_bterm_groups.push_back(std::move(group));
 }
 
+void dbBlock::setBTermTopLayerGrid(dbTechLayer* layer,
+                                   int x_step,
+                                   int y_step,
+                                   Rect region,
+                                   int pin_width,
+                                   int pin_height,
+                                   int keepout)
+{
+  _dbBlock* block = (_dbBlock*) this;
+  _dbBTermTopLayerGrid& top_layer_grid = block->_bterm_top_layer_grid;
+
+  utl::Logger* logger = block->getImpl()->getLogger();
+  const odb::Rect& die_area = getDieArea();
+
+  if (!die_area.contains(region)) {
+    logger->error(
+        utl::ODB, 124, "Top layer grid region is out of the die area.");
+  }
+
+  top_layer_grid.layer = layer->getId();
+  top_layer_grid.x_step = x_step;
+  top_layer_grid.y_step = y_step;
+  top_layer_grid.region = region;
+  top_layer_grid.pin_width = pin_width;
+  top_layer_grid.pin_height = pin_height;
+  top_layer_grid.keepout = keepout;
+}
+
+dbBlock::dbBTermTopLayerGrid dbBlock::getBTermTopLayerGrid()
+{
+  _dbBlock* block = (_dbBlock*) this;
+
+  dbBlock::dbBTermTopLayerGrid top_layer_grid;
+
+  odb::dbTech* tech = getDb()->getTech();
+  dbTechLayer* layer = nullptr;
+  const auto& top_grid = block->_bterm_top_layer_grid;
+
+  if (top_grid.layer.isValid()) {
```

Comment:
When would it not be valid?  `top_layer_grid.layer = layer->getId();` is going to fail if layer is null (which should be checked and error'ed).

Reply:
Insided PPL we are calling the `getBTermTopLayerGrid` function even when the top layer grid was not defined. Then it checks if the layer is `nullptr` to detect if it was defined or not. I'll change this logic on PPL to avoid calling this function when we don't have the top layer grid defined.